### ISSUE TITLE
RNN in PyTorch

### ISSUE THREAD

**Issue by ketankokane94** [amnag94/Accident-Prediction#3]
Check for the implementation of RNN models in Pytorch.

This model does not have to deal with actual image data, just the features extracted from the VGG-16. 

So this model needs to know the exact output of the VGG-16 model.

**Comment by ketankokane94** [comment]
https://pytorch.org/tutorials/intermediate/char_rnn_classification_tutorial.html

will be using this as reference

**Comment by ketankokane94** [comment]
Decide upon the hyperparameters, **easy ones like learning rate, the number of epochs**.
and also the **difficult ones, like the number of hidden layers** number of output neurons(which can also be just one, as we are predicting the probability)

**Comment by ketankokane94** [comment]
Each loop of training will:

1. Create input and target tensors
2. Create a zeroed initial hidden state
3. Read each frame in and
4. Keep hidden state for frame letter
5. Compare final output to ground truth
6. Back-propagate
7. Return the output and loss

**Comment by ketankokane94** [comment]
```python
epochs = 100
current_loss = 0
all_losses = []

for iter in range(1, epochs + 1):
    true_value_tensor, video_sequence_tensor = randomTrainingExample()
    output, loss = train(true_value_tensor, video_sequence_tensor)
    current_loss += loss
```

**Comment by ketankokane94** [comment]
```python
learning_rate = 0.005 
# If set too high, it might explode. If too low, it might not learn

def train(true_value_tensor, video_sequence_tensor):
    hidden = rnn.initHidden()

    rnn.zero_grad()

    for i in range(video_sequence_tensor.size()[0]): 
# for every frame in video sequence
        prediction_tensor, hidden = rnn(video_sequence_tensor[i], hidden) 
# get prediction for every frame

    loss = criterion(prediction_tensor, true_value_tensor) 
# we want Exponential Loss here
    loss.backward() # backpropogate

    # Add parameters' gradients to their values, multiplied by learning rate
    for p in rnn.parameters():
        p.data.add_(-learning_rate, p.grad.data)

    return prediction_tensor, loss.item() # return final predicted and total loss
```

**Comment by ketankokane94** [comment]
```python
import numpy as np

number_of_frame = 13
features_of_video_Seq = []

for _ in range(number_of_frame):
    features_of_video_Seq.append(
        np.random.rand(1 , 4096).flatten() # this array would be generated by VGG
    )

features_of_video_Seq = np.array(features_of_video_Seq)
video_tensor = torch.from_numpy(features_of_video_Seq)
print('video_tensor.shape', video_tensor.shape)

for frame_idx in range(video_tensor.shape[0]):
    print('frame_tensor.shape', video_tensor[frame_idx].shape)
```

video_tensor.shape torch.Size([13, 4096])
frame_tensor.shape torch.Size([4096])
frame_tensor.shape torch.Size([4096])
frame_tensor.shape torch.Size([4096])
frame_tensor.shape torch.Size([4096])
frame_tensor.shape torch.Size([4096])
frame_tensor.shape torch.Size([4096])
frame_tensor.shape torch.Size([4096])
frame_tensor.shape torch.Size([4096])
frame_tensor.shape torch.Size([4096])
frame_tensor.shape torch.Size([4096])
frame_tensor.shape torch.Size([4096])
frame_tensor.shape torch.Size([4096])
frame_tensor.shape torch.Size([4096])
